docs/preview/04-Guidance/migrate-from-v2-to-v3.md

```diff
@@ -0,0 +1,228 @@
+* 🗑️ `CriticalExceptionFilter`
+* 🗑️ `SecretStoreBuilder.AddCriticalException<...>(...)`
+
+Custom secret providers can use the `SecretResult` model in case of interrupted/non-transient failures.
+
+### 🚛 Move caching from providers to store
+Caching is centralized on the secret store instead of spread across secret providers. Just as before v3, caching happens internally, only now custom secret provider are registered without any mentioning of secret caching.
+
+This affects the existing caching-types and (extension) members, which are removed/unavailable in v3:
+* 🗑️ `(I)CachedSecretProvider`
+* 🗑️ `(I)CacheConfiguration`
+* 🗑️ `.WithCaching(...)` on secret provider
+* 🗑️ `.GetCachedProvider(...)` on secret store.
+
+```diff
+services.AddSecretStore(store =>
+{
+    var cacheDuration = TimeSpan.FromMinutes(5);
+-   store.AddProvider(new MySecretProvider().WithCaching(cacheDuration));
++   store.AddProvider(new MySecretProvider());
++   store.UseCaching(cacheDuration);
+});
+```
+
+Ignoring the cache at secret retrieval-time can still be done via a new method overload on the `ISecretStore` instead on the secret provider itself.
+
+```diff
+- using Arcus.Security.Core.Caching;
+- using Arcus.Security.Core;
++ using Arcus.Security;
+
+ISecretStore store = ...
+
+- ICachedSecretProvider provider = store.GetCachedProvider("Admin secrets");
+- Secret secret = await provider.GetSecretAsync("<name>", ignoreCache: true);
++ SecretResult result = await store.GetSecretAsync("<name>", options =>
++ {
++     options.UseCache = false;
++ });
+```
+
+Invalidating a secret within a custom secret provider now happens via the `ISecretStoreContext` that can be passed upon creating the provider.
+
+```diff
+- public class MySecretProvider : ICachedSecretProvider
++ public class MySecretProvider(ISecretStoreContext context) : ISecretProvider
+{
+    public async Task SetSecretAsync(string name, string value)
+    {
+        // Implementation omitted.
+
+-       await InvalidateSecretAsync(name);
++       await context.Cache.InvalidateSecretAsync(name);
+    }
+
+-   public Task InvalidateSecretAsync(string secretName) { ... }
+
+    // Implementation omitted.
+}
+```
+
+### 🗑️ Removed `GetRawSecret*` overloads
+Starting from v3, there is no distinction anymore between 'secrets' and 'raw secrets' (meaning: directly accessing the secret's value). All secret interactions happen via the asynchronous/synchronous `GetSecret(Async)` methods on the `ISecretStore`.
+
+:::tip[implicit overload on `SecretResult`]
+There exists an `implicit operator` overload on the `SecretResult`, which means that that secret retrievals can also be written without checking for failures.
+```csharp
+string secretValue = await store.GetSecretAsync("<name>");
+```
+Just be aware that in case of a failure, an exception will still be thrown.
```

Comment:
Maybe we should've implemented that in such a way that we return `null` instead of throwing an exception ?

Reply:
Might be better communicated to users that it is only there if there is a result, otherwise we could have succes with null or not found being rather similar.
Also following the built in result type in F# here as well.